src/stan/io/reader.hpp

```diff
@@ -1227,116 +1317,351 @@ class reader {
+   * @param vec_c columns corresponding the nonzero entries.
+   * @param n Number of rows.
+   * @param m Number of columns.
+   * @param lp The log probability
+   * @return Eigen::SparseMatrix made up of the next scalars.
+   */
+  template <typename TL, typename VecR, typename VecC,
+            typename = require_all_std_vector_vt<std::is_integral, VecR, VecC>>
+  inline sparse_matrix_t sparse_matrix_lb_constrain(const TL lb,
+                                                    const VecR& vec_r,
+                                                    const VecC& vec_c, size_t n,
+                                                    size_t m, T& lp) {
+    stan::math::check_matching_sizes("sparse_matrix_lb_constrain", "vec_r",
+                                     vec_r, "vec_c", vec_c);
+    Eigen::SparseMatrix<T> ret_mat(n, m);
+    if (m == 0 || n == 0) {
+      return ret_mat;
+    }
+    using triplet_type = Eigen::Triplet<T>;
+    std::vector<triplet_type> triplet_list(vec_r.size());
+    for (auto i = 0; i < vec_r.size(); i++) {
+      triplet_list.emplace_back(
+          triplet_type(vec_r[i], vec_c[i], scalar_lb_constrain(lb, lp)));
+    }
+    ret_mat.setFromTriplets(triplet_list.begin(), triplet_list.end());
+    return ret_mat;
+  }
+
   template <typename TU>
   inline vector_t vector_ub(const TU ub, size_t m) {
     vector_t v(m);
-    for (size_t i = 0; i < m; ++i)
+    for (size_t i = 0; i < m; ++i) {
       v(i) = scalar_ub(ub);
+    }
     return v;
   }
 
   template <typename TU>
   inline vector_t vector_ub_constrain(const TU ub, size_t m) {
     vector_t v(m);
-    for (size_t i = 0; i < m; ++i)
+    for (size_t i = 0; i < m; ++i) {
       v(i) = scalar_ub_constrain(ub);
+    }
     return v;
   }
 
   template <typename TU>
-  inline vector_t vector_ub_constrain(const TU ub, size_t m, T &lp) {
+  inline vector_t vector_ub_constrain(const TU ub, size_t m, T& lp) {
     vector_t v(m);
-    for (size_t i = 0; i < m; ++i)
+    for (size_t i = 0; i < m; ++i) {
       v(i) = scalar_ub_constrain(ub, lp);
+    }
     return v;
   }
 
   template <typename TU>
   inline row_vector_t row_vector_ub(const TU ub, size_t m) {
     row_vector_t v(m);
-    for (size_t i = 0; i < m; ++i)
+    for (size_t i = 0; i < m; ++i) {
       v(i) = scalar_ub(ub);
+    }
     return v;
   }
 
   template <typename TU>
   inline row_vector_t row_vector_ub_constrain(const TU ub, size_t m) {
     row_vector_t v(m);
-    for (size_t i = 0; i < m; ++i)
+    for (size_t i = 0; i < m; ++i) {
       v(i) = scalar_ub_constrain(ub);
+    }
     return v;
   }
 
   template <typename TU>
-  inline row_vector_t row_vector_ub_constrain(const TU ub, size_t m, T &lp) {
+  inline row_vector_t row_vector_ub_constrain(const TU ub, size_t m, T& lp) {
     row_vector_t v(m);
-    for (size_t i = 0; i < m; ++i)
+    for (size_t i = 0; i < m; ++i) {
       v(i) = scalar_ub_constrain(ub, lp);
+    }
     return v;
```

Comment:
[very optional]
I really want to go higher order and just turn all these into generic comprehensions.  Not sure it's a good idea, though.  Something like:

```
template <typename M, class F>
comprehend_matrix(int m, int n, const F& f) {
  M x(m, n);
  for (int i = 0; i < x.size(); ++i) x(i) = f();
  return M;
}
```